### ISSUE TITLE
APIs for CSV import functionality

### ISSUE THREAD

**Issue by kgodey** [centerofci/mathesar#98]
**Problem**

Our current CSV import functionality was meant to be temporary and has a few issues:
- It's synchronous and blocks response until the CSV file is imported.
- It does not save the CSV file.
- It's built using Django Forms and cannot be used by the Svelte-based frontend.

**Proposed solution**

CSV imports should be possible using the API. There should be separate endpoints for the following tasks:
- Uploading a CSV file and saving it to disk.
- Showing the user a preview of the CSV file (with types inferred although that's not in scope for this issue and will need to wait until #93 is done)
- Showing the status of the CSV upload and type inference.
- Actually creating a table (eventually with types confirmed by the user, once #93 is done)

**Comment by mathemancer** [comment]
I'm not sure that the 'show the user a preview' endpoint should know or care if the types are inferred.  Shouldn't that just be an endpoint to look at a preview of _any_ table?  Or does this mean a separate endpoint to show part of an actual CSV file?  I'd advocate only showing data from tables in the DB (even if the type info is cached somewhere else until it's finalized).

**Comment by kgodey** [comment]
@mathemancer I'm not sure where the type information is going to come from, but we need to show types in this endpoint because the user needs to see them next to the previewed columns and correct the types if needed. Here's the part of the low fidelity wireframes corresponding to the functionality:

<img width="837" alt="Screen Shot 2021-05-03 at 2 16 16 PM" src="https://user-images.githubusercontent.com/287034/116915613-3d269d00-ac1a-11eb-96fa-1c4f2f9e533c.png">

**Comment by mathemancer** [comment]
Sure, but my point is we need to show types in the normal "view table" endpoint, so why not just call that with the name of the table holding the data with the inferred types on the columns (limited as appropriate for a preview).  I.e., the post to "create table" would just return the table name (or id), and then a call with that with the appropriate limit would have all the required data.  "Finish import" wouldn't really be a change on the database level, but maybe could flag the table as "stored" or something in the model maybe?

**Comment by kgodey** [comment]
@mathemancer Sure, that works, I hadn't figured out any implementation details yet, I was just documenting the requirements for the UI.

We should finish our discussion about how to implement type inference from the meeting today so we're on the same page. Do you want to start a GitHub discussion for that?

**Comment by mathemancer** [comment]
I've started a discussion:  https://github.com/centerofci/mathesar/discussions/104
